econml/cate_estimator.py

```diff
@@ -5,63 +5,47 @@
+from functools import wraps
+from copy import deepcopy
+from warnings import warn
 from .bootstrap import BootstrapEstimator
-from .inference import BootstrapOptions
+from .inference import BootstrapInference
 from .utilities import tensordot, ndim, reshape, shape
 
 
 class BaseCateEstimator(metaclass=abc.ABCMeta):
-    """
-    Base class for all CATE estimators in this package.
+    """Base class for all CATE estimators in this package."""
 
-    Parameters
-    ----------
-    inference: string, inference method, or None
-        Method for performing inference.  All estimators support 'bootstrap'
-        (or an instance of `BootstrapOptions`), some support other methods as well.
-
-    """
-
-    _inference_options = {'bootstrap': BootstrapOptions()}
-    _bootstrap_whitelist = {'effect', 'marginal_effect'}
-
-    @abc.abstractmethod
-    def __init__(self, inference):
+    def _get_inference_options(self):
         """
-        Initialize the estimator.
-
-        All subclass overrides should complete with a call to this method on the super class,
-        since it enables bootstrapping.
+        Produce a dictionary mapping string names to `Inference` types.
 
+        This is used by the `fit` method when a string is passed rather than an `Inference` type.
         """
-        if inference in self._inference_options:
-            inference = self._inference_options[inference]
-        if isinstance(inference, BootstrapOptions):
-            # Note that fit (and other methods) check for the presence of a _bootstrap attribute
-            # to determine whether to delegate to that object or not;
-            # The clones wrapped inside the BootstrapEstimator will not have that attribute since
-            # it's assigned *after* creating the estimator
-            self._bootstrap = BootstrapEstimator(self, inference.n_bootstrap_samples, inference.n_jobs)
-        self._inference = inference
-
-    def __getattr__(self, name):
-        suffix = '_interval'
-        if name.endswith(suffix) and name[: - len(suffix)] in self._bootstrap_whitelist:
-            if hasattr(self, '_bootstrap'):
-                return getattr(self._bootstrap, name)
+        return {'bootstrap': BootstrapInference}
+
+    def _get_inference(self, inference):
+        options = self._get_inference_options()
+        if isinstance(inference, str):
+            if inference in options:
+                inference = options[inference]()
             else:
-                raise AttributeError('\'%s\' object does not support attribute \'%s\'; '
-                                     'consider passing inference=\'bootstrap\' when initializing'
-                                     % (type(self).__name__, name))
-        else:
-            raise AttributeError('\'%s\' object has no attribute \'%s\''
-                                 % (type(self).__name__, name))
+                raise ArgumentError("Inference option '%s' not recognized; valid values are %s" %
```

Comment:
Argument error is not defined in import

Reply:
Fixed, thanks (should have been `ValueError`).